Add stage3_mlxupdate image build & upload to .travis.yml

stephen-soltesz opened the pull request:
This change adds a new build and deploy step to the epoxy-images .travis.yaml config.

This change adds support for installing the gcloud CLI tools in order to fetch a cached version of the mft archive (~510MB) faster than it can be downloaded from mellanox. The resulting kernel and initram images are copied to `gs://epoxy-mlab-staging/stage3_mlxupdate/` bucket.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/epoxy-images/19)
<!-- Reviewable:end -->

gfr10598 commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[.travis.yml, line 55 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34fXFjuFPQ2IGkTOJ6:-L34fXFjuFPQ2IGkTOJ7:b2tgmc6) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L55)):*
> ```YAML
> # Build stage3 mlxupdate image.
> #
> # Note: download the mellanox firmware tools from a faster private bucket.
> ```

This comment reads like it is a change from something else.  Instead, something like "download ... from cache in private bucket, since origin (http://...) is slow.

---

*[.travis.yml, line 56 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34fzTYiKnbKcNjleQV:-L34fzTYiKnbKcNjleQW:bre2q9d) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L56)):*
> ```YAML
> #
> # Note: download the mellanox firmware tools from a faster private bucket.
> - gsutil cp gs://vendor-mlab-oti/epoxy-images/mft-4.4.0-44.tgz $TRAVIS_BUILD_DIR/
> ```

Suggest this should be vendor-mlab-oti/mellanox/...

---

*[.travis.yml, line 58 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34hB4chAxf5XArEik8:-L34hB4chAxf5XArEik9:b-rgbj37) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L58)):*
> ```YAML
> - gsutil cp gs://vendor-mlab-oti/epoxy-images/mft-4.4.0-44.tgz $TRAVIS_BUILD_DIR/
> 
> # Note: build must be privilged to mount /proc & /sys during debootstrap build.
> ```

typo - privileged

---

*[.travis.yml, line 63 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34h2bMhuP-W22TBaXR:-L34h2bMhuP-W22TBaXS:b-pz3oag) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L63)):*
> ```YAML
> #
> # TODO: add support for create_update_iso.sh. Possibly call from setup_stage1.sh
> - time docker run --privileged -v $PWD:/images epoxy-images-builder
> ```

is $PWD better than $TRAVIS_BUILD_DIR here?

---

*[.travis.yml, line 72 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34hgVm3DdOHYUH-IDz:-L34hgVm3DdOHYUH-IE-:b-ewtd7r) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L72)):*
> ```YAML
> # after a merge with matching "on:" conditions.
> deploy:
> # SANDBOX: before code review for development code in a specific branch.
> ```

Could you add one or two lines indicating what is deployed where, both here and below?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy-images/19)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-:-L34hrRzGBSyUXW0-CWb:bnfp4nl)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Thank you for the review!

---

Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[.travis.yml, line 55 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34fXFjuFPQ2IGkTOJ6:-L34rH2jk6-Aoo8ygNFS:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L55)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

This comment reads like it is a change from something else.  Instead, something like "download ... from cache in private bucket, since origin (http://...) is slow.
</blockquote></details>

Done.

---

*[.travis.yml, line 56 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34fzTYiKnbKcNjleQV:-L34oISiZyTLk6Xw_b7d:bxjw5k6) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L56)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Suggest this should be vendor-mlab-oti/mellanox/...
</blockquote></details>

My thinking with the current name is that it provides a back-reference to the repo that depends on it. That only works as long as there's only one reference. But, right now there is only one reference expected. So, I've added a README to that bucket explaining the origin of this file that spells out the fact that "mft" means Mellanox Firmware Tools. That is a necessary hint.

---

*[.travis.yml, line 58 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34hB4chAxf5XArEik8:-L34mcryPcCeiQ6gDb7h:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L58)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

typo - privileged
</blockquote></details>

Done.

---

*[.travis.yml, line 63 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34h2bMhuP-W22TBaXR:-L34mjso2nyNWk4TEL6Y:b8862hv) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L63)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

is $PWD better than $TRAVIS_BUILD_DIR here?
</blockquote></details>

Using a consistent path is better, so I've changed it to use `TRAVIS_BUILD_DIR` throughout.

The original choice was really only a consequence of me running this on my workstation first and then copying the command into travis.yml.

---

*[.travis.yml, line 72 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34hgVm3DdOHYUH-IDz:-L34n_f_Dnv7R_XkIodE:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L72)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Could you add one or two lines indicating what is deployed where, both here and below?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-:-L34rbIz9IxzlCmFu-mt:b-rrkdbt)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[.travis.yml, line 56 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34fzTYiKnbKcNjleQV:-L34wxI4HZtq6Zr0k3TN:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L56)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

My thinking with the current name is that it provides a back-reference to the repo that depends on it. That only works as long as there's only one reference. But, right now there is only one reference expected. So, I've added a README to that bucket explaining the origin of this file that spells out the fact that "mft" means Mellanox Firmware Tools. That is a necessary hint.
</blockquote></details>

Done.

---

*[.travis.yml, line 63 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34h2bMhuP-W22TBaXR:-L34wwl63SsJX7erFbDP:b-896fix) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L63)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Using a consistent path is better, so I've changed it to use `TRAVIS_BUILD_DIR` throughout.

The original choice was really only a consequence of me running this on my workstation first and then copying the command into travis.yml.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy-images/19)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
*[.travis.yml, line 56 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34fzTYiKnbKcNjleQV:-L34wyHkfuVVcG1rG8Rs:behunog) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L56)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Ok - I'll see if I can adopt that perspective.  Not sure it makes best sense.  If two repos both depend on one artifact, we would have two copies of the artifact, I guess.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy-images/19)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[.travis.yml, line 56 at r1](https://reviewable.io:443/reviews/m-lab/epoxy-images/19#-L34fzTYiKnbKcNjleQV:-L38nW0VAg0GgwqNFoU2:b-wka88) ([raw file](https://github.com/m-lab/epoxy-images/blob/c185839ec0f1a048ce3df1888fbd820948fb522b/.travis.yml#L56)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

Ok - I'll see if I can adopt that perspective.  Not sure it makes best sense.  If two repos both depend on one artifact, we would have two copies of the artifact, I guess.
</blockquote></details>

My thought is that determining the set of references can be expensive without hints at the source. And, storage is relatively much less expensive.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/epoxy-images/19)*
<!-- Sent from Reviewable.io -->
